src/core/server/saved_objects/import/check_conflicts.ts

```diff
@@ -0,0 +1,220 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
+
+import { v4 as uuidv4 } from 'uuid';
+import {
+  SavedObject,
+  SavedObjectsClientContract,
+  SavedObjectsImportError,
+  SavedObjectsImportRetry,
+} from '../types';
+import { ISavedObjectTypeRegistry } from '..';
+
+interface CheckConflictsOptions {
+  savedObjectsClient: SavedObjectsClientContract;
+  typeRegistry: ISavedObjectTypeRegistry;
+  namespace?: string;
+}
+
+interface GetImportIdMapForRetriesOptions {
+  typeRegistry: ISavedObjectTypeRegistry;
+  retries: SavedObjectsImportRetry[];
+}
+
+type InexactMatch<T> = {
+  object: SavedObject<T>;
+  destinations: Array<{ id: string; title?: string; updatedAt?: string }>;
+};
+type Left<T> = { tag: 'left'; value: InexactMatch<T> };
+type Right<T> = { tag: 'right'; value: SavedObject<T> };
+type Either<T> = Left<T> | Right<T>;
+const isLeft = <T>(object: Either<T>): object is Left<T> => object.tag === 'left';
+
+const createQueryTerm = (input: string) => input.replace(/\\/g, '\\\\').replace(/\"/g, '\\"');
+const createQuery = (type: string, id: string, rawIdPrefix: string) =>
+  `"${createQueryTerm(`${rawIdPrefix}${type}:${id}`)}" | "${createQueryTerm(id)}"`;
+const getAmbiguousConflicts = (objects: Array<SavedObject<{ title?: string }>>) =>
```

Comment:
Here's the diff that I'm about to push:
```
-const getAmbiguousConflicts = (objects: Array<SavedObject<{ title?: string }>>) =>
+const transformObjectsToAmbiguousConflictFields = (
+  objects: Array<SavedObject<{ title?: string }>>
+) =>
   objects
     .map(({ id, attributes, updated_at: updatedAt }) => ({
       id,
       title: attributes?.title,
       updatedAt,
     }))
+    // Sort for two reasons: 1. consumers may want to identify multiple errors that have the same sources (by stringifying the `sources`
+    // array of each object they can be compared), and 2. it will be a less confusing experience for end-users if several ambiguous
+    // conflicts that share the same destinations all show those destinations in the same order.
     .sort((a, b) => (a.id > b.id ? 1 : b.id > a.id ? -1 : 0));
```

Sorting by `id` was a bit of an arbitrary choice. Perhaps it makes more sense to sort by `updatedAt` in descending order, and then by `id`? That way the selection box that the end user sees will always have the last-modified result at the top.

Reply:
> Sorting by id was a bit of an arbitrary choice. Perhaps it makes more sense to sort by updatedAt in descending order, and then by id? That way the selection box that the end user sees will always have the last-modified result at the top.

Sorting by `updatedAt` makes sense to me. The secondary sort may not be necessary, but it can't hurt either (how likely/important is it that two objects are updated at the same exact time, where the secondary sort by ID would help the end-user make a decision?).

My main comment was that it wasn't clear to me why the sort was taking place. I didn't realize the sort we performed here when getting the ambiguous conflicts influenced the end result we displayed in the UI

Reply:
After the second round of reviews is complete, I'll change this to sort by `updatedAt`.
The first reason (consumers may want to identify multiple errors that have the same sources) is no longer possible since we handle Ambiguous Source conflicts differently than first envisioned. I'll update the comment as well.